File: java/robot-name/src/main/java/Robot.java
public class Robot {
    private String name = initialiseName();

    public Robot() {
    }

    public String getName() {
        return name;
    }

    public void reset() {
        name = initialiseName();
    }

    private String initialiseName() {
        return String.format("%c%c%d%d%d",
                upperCaseLetter(),
                upperCaseLetter(),
                digit(),
                digit(),
                digit()
        );
    }

    private Character upperCaseLetter() {
        int firstAsciiUpperCaseLetterCode = 65;
        int totalNumberOfLetters = 26;
        return (char) (firstAsciiUpperCaseLetterCode + (Math.random() * totalNumberOfLetters));
    }

    private Integer digit() {
        int maximumDigit = 9;
        int zeroBasedAdjustment = 1;
        return (int) (Math.random() * (maximumDigit + zeroBasedAdjustment));
    }
}

File: kotlin/robot-name/src/main/kotlin/Robot.kt
class Robot {
    val letters = ('A'..'Z').toList()
    val numbers = (0..9).toList()
    var previousNames = mutableSetOf<String>()
    var name = generateName()

    private fun generateName(): String {
        val proposedName = letter() + letter() + number() + number() + number()

        if (previousNames.contains(proposedName)) {
            return generateName()
        }
        previousNames.add(proposedName)
        return proposedName
    }

    private fun letter() = letters[(Math.random() * letters.size).toInt()].toString()

    private fun number() = numbers[(Math.random() * numbers.size).toInt()].toString()

    fun reset() {
        name = generateName()
    }
}
